src/core/writer.js

```diff
@@ -25,36 +25,68 @@ import { SimpleDOMNode, SimpleXMLParser } from "./xml_parser.js";
 import { BaseStream } from "./base_stream.js";
 import { calculateMD5 } from "./crypto.js";
 
-function writeObject(ref, obj, buffer, transform) {
+async function writeObject(ref, obj, buffer, transform) {
   buffer.push(`${ref.num} ${ref.gen} obj\n`);
   if (obj instanceof Dict) {
-    writeDict(obj, buffer, transform);
+    await writeDict(obj, buffer, transform);
   } else if (obj instanceof BaseStream) {
-    writeStream(obj, buffer, transform);
+    await writeStream(obj, buffer, transform);
   }
   buffer.push("\nendobj\n");
 }
 
-function writeDict(dict, buffer, transform) {
+async function writeDict(dict, buffer, transform) {
   buffer.push("<<");
   for (const key of dict.getKeys()) {
     buffer.push(` /${escapePDFName(key)} `);
-    writeValue(dict.getRaw(key), buffer, transform);
+    await writeValue(dict.getRaw(key), buffer, transform);
   }
   buffer.push(">>");
 }
 
-function writeStream(stream, buffer, transform) {
+async function writeStream(stream, buffer, transform) {
   let string = stream.getString();
   if (transform !== null) {
     string = transform.encryptString(string);
   }
+
+  const filter = stream.dict.getRaw("Filter");
+  const flateDecode = Name.get("FlateDecode");
+  const MIN_LENGTH_FOR_COMPRESSING = 192;
+  let mustCompress = string.length >= MIN_LENGTH_FOR_COMPRESSING;
+
+  if (Array.isArray(filter)) {
+    if (filter.includes(flateDecode)) {
+      mustCompress = true;
+    } else if (mustCompress) {
+      filter.push(flateDecode);
+    }
+  } else if (mustCompress) {
+    if (!filter) {
+      stream.dict.set("Filter", flateDecode);
+    } else if (filter !== flateDecode) {
+      stream.dict.set("Filter", [filter, flateDecode]);
+    }
+  }
+
+  // eslint-disable-next-line no-undef
+  if (CompressionStream && mustCompress) {
+    const byteArray = stringToBytes(string);
+    // eslint-disable-next-line no-undef
+    const cs = new CompressionStream("deflate");
+    const writer = cs.writable.getWriter();
+    writer.write(byteArray);
+    writer.close();
+    const compressedData = await new Response(cs.readable).arrayBuffer();
+    string = bytesToString(new Uint8Array(compressedData));
```

Comment:
I'm assuming that the fact that it's documented as returning UTF-8 encoded data, according to https://developer.mozilla.org/en-US/docs/Web/API/Response/text, means that we can't simply do the following?
```suggestion
    string = await new Response(cs.readable).text();
```